[1.21] Allow looking up a Dimensions translated name

ApexModder opened the pull request:
This PR introduces a couple of new methods to `Level`, allowing for the lookup of the translation key and translated component for a given dimension. If no translation currently exists, the translated component falls back to the registry name.

The format for these translations is `dimension.<namespace>.<path>`. For example, the overworld would be `dimension.minecraft.overworld`.

Additionally, a new method has been added to `LanguageProvider` to facilitate data generation of translations for dimensions. This method takes in the dimension’s registry key and translated name.

ApexModder committed:
Initial implementation

ApexModder committed:
Add simple javadocs

ApexModder committed:
Fix typo in javadoc

XFactHD commented on src/main/java/net/neoforged/neoforge/common/data/LanguageProvider.java:
```diff
@@ -133,4 +136,8 @@ public void add(String key, String value) {
             throw new IllegalStateException("Duplicate translation key " + key);
     }
+
+    public void add(ResourceKey<Level> dimension, String value) {
```
This should have a distinct name (i.e. `addDimension()`) as overloads which only change generics are not valid

ApexModder replied:
👍

ApexModder committed:
Rename method to addDimension

Technici4n commented on src/main/java/net/neoforged/neoforge/server/command/DimensionsCommand.java:
```diff
@@ -28,13 +28,13 @@ class DimensionsCommand {
 
                     types.keySet().stream().sorted().forEach(key -> {
-                        ctx.getSource().sendSuccess(() -> Component.literal(key + ": " + types.get(key).stream().map(ResourceLocation::toString).sorted().collect(Collectors.joining(", "))), false);
+                        ctx.getSource().sendSuccess(() -> Component.literal(key + ": " + types.get(key).stream().map(Component::getString).sorted().collect(Collectors.joining(", "))), false);
```
Not blocking: This discard formatting, right? Is there no better way to combine multiple components?

ApexModder replied:
Expanded out in latest commit, now uses `component.append` to not resolve translations server side.

XFactHD commented on src/main/java/net/neoforged/neoforge/common/data/LanguageProvider.java:
```diff
@@ -114,11 +115,15 @@ public void addTag(Supplier<? extends TagKey<?>> key, String name) {
     }
+
+    public void addDimension(ResourceKey<Level> dimension, String value) {
+        add(dimension.location().toLanguageKey(net.neoforged.neoforge.common.extensions.ILevelExtension.TRANSLATION_PREFIX), value);
```
```suggestion
        add(dimension.location().toLanguageKey(ILevelExtension.TRANSLATION_PREFIX), value);
```
Add an import instead of using an FQN

ApexModder replied:
Noticed this right after I clicked the push button, should be fixed now

ApexModder replied:
Wait no... My mistake again, the name `LanguageProvider` made me think its a vanilla file, so I FQN'd but its a neo file 🤦.

This is what I get for doing a "quick rebase" immediately after waking up.

Edit: Should be fixed in latest push

XFactHD commented on src/main/java/net/neoforged/neoforge/common/data/LanguageProvider.java:
```diff
@@ -114,11 +115,15 @@ public void addTag(Supplier<? extends TagKey<?>> key, String name) {
 
     public void add(TagKey<?> tagKey, String name) {
-        add(Tags.getTagTranslationKey(tagKey), name);
+        add(net.neoforged.neoforge.common.Tags.getTagTranslationKey(tagKey), name);
```
Why is this being changed from an import to an FQN?

ApexModder replied:
I might have messed up the original rebase but should be fixed now

ApexModder committed:
Apply formatting

Third times the charm 🤞

ApexModder committed:
Fix up dimension command

Should now actually make use of translations client side, getString resolved server side

XFactHD committed:
Merge branch '1.21.x' into pr/dimension-translations

Technici4n reviewed (approved):
LGTM
